Net buffer macros rely on GCC extension

With the introduction of "flexible user data" in net buffers, the implementation apparently relies on a GCC specific extension.
The following warning is seen:

```
buf.c:38:1: warning: field 'b' with variable sized type 'struct net_buf' not at the end of a struct or class is a GNU extension [-Wgnu-variable-sized-type-not-at-end]
NET_BUF_POOL_FIXED_DEFINE(num_complete_pool, 1, NUM_COMLETE_EVENT_SIZE, 8, NULL);
^
```
This in turn causes linker errors.

The offending commits are more specifically:
> df327eeb58: net: buf: `POOL_FIXED_DEFINE` explicit user data
> bb780eff99: net: buf: move `user_data` to flexible arr member

The Zephyr code base should not rely on any specific tool-chain.

**Impact**
Unable to link proprietary firmware after introduction.

--- a/include/net/buf.h
+++ b/include/net/buf.h
@@ -1045,16 +1045,17 @@ struct net_buf_pool {
 	}
 #endif /* CONFIG_NET_BUF_POOL_USAGE */
 
-#define _NET_BUF_ARRAY_DEFINE(_name, _count, _ud_size)                                    \
-	struct _net_buf_##_name {struct net_buf b; uint8_t ud[_ud_size]; };               \
-	BUILD_ASSERT(_ud_size <= UINT8_MAX);                                              \
-	BUILD_ASSERT(offsetof(struct net_buf, user_data) ==                               \
-		offsetof(struct _net_buf_##_name, ud), "Invalid offset");                 \
-	BUILD_ASSERT(__alignof__(struct net_buf) ==                                       \
-		__alignof__(struct _net_buf_##_name), "Invalid alignment");               \
-	BUILD_ASSERT(sizeof(struct _net_buf_##_name) ==                                   \
-		ROUND_UP(sizeof(struct net_buf) + _ud_size, __alignof__(struct net_buf)), \
-		"Size cannot be determined");                                             \
+#define _NET_BUF_ARRAY_DEFINE(_name, _count, _ud_size)					       \
+	struct _net_buf_##_name { uint8_t b[sizeof(struct net_buf)];			       \
+				  uint8_t ud[_ud_size]; } __net_buf_align;		       \
+	BUILD_ASSERT(_ud_size <= UINT8_MAX);						       \
+	BUILD_ASSERT(offsetof(struct net_buf, user_data) ==				       \
+		     offsetof(struct _net_buf_##_name, ud), "Invalid offset");		       \
+	BUILD_ASSERT(__alignof__(struct net_buf) ==					       \
+		     __alignof__(struct _net_buf_##_name), "Invalid alignment");	       \
+	BUILD_ASSERT(sizeof(struct _net_buf_##_name) ==					       \
+		     ROUND_UP(sizeof(struct net_buf) + _ud_size, __alignof__(struct net_buf)), \
+		     "Size cannot be determined");					       \
 	static struct _net_buf_##_name _net_buf_##_name[_count] __noinit
 
 extern const struct net_buf_data_alloc net_buf_heap_alloc;

